### PR TITLE
Add JS and Protobuf.js benchmark, fix js's reader.skipGroup

### PR DESCRIPTION
**Pull request by BSBandme**
Protobuf.js performance:
```

ProtobufJs Benchmark: 
benchmarking message benchmarks.proto2.GoogleMessage2 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/datasets/google_message2/dataset.google_message2.pb's performance ...

protobuf.js static decoding x 2,618 ops/sec ±1.23% (90 runs sampled)
protobuf.js static encoding x 782 ops/sec ±1.16% (91 runs sampled)
Throughput for decoding: 208.55442550417717MB/s
Throughput for encoding: 62.331349090264304MB/s

benchmarking message benchmarks.proto3.GoogleMessage1 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/datasets/google_message1/proto3/dataset.google_message1_proto3.pb's performance ...

protobuf.js static decoding x 526,123 ops/sec ±0.66% (88 runs sampled)
protobuf.js static encoding x 192,045 ops/sec ±1.30% (92 runs sampled)
Throughput for decoding: 113.64940927321399MB/s
Throughput for encoding: 41.2206793225958MB/s

benchmarking message benchmarks.proto2.GoogleMessage1 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/datasets/google_message1/proto2/dataset.google_message1_proto2.pb's performance ...

protobuf.js static decoding x 514,598 ops/sec ±0.34% (93 runs sampled)
protobuf.js static encoding x 198,119 ops/sec ±1.66% (90 runs sampled)
Throughput for decoding: 111.50854424334813MB/s
Throughput for encoding: 42.376785111378936MB/s

benchmarking message benchmarks.google_message4.GoogleMessage4 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message4.pb's performance ...

protobuf.js static decoding x 4.00 ops/sec ±13.52% (15 runs sampled)
protobuf.js static encoding x 2.66 ops/sec ±2.70% (11 runs sampled)
Throughput for decoding: 254.4844855864752MB/s
Throughput for encoding: 187.23144405093612MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_4.pb's performance ...

protobuf.js static decoding x 748 ops/sec ±3.19% (91 runs sampled)
protobuf.js static encoding x 70.24 ops/sec ±3.04% (60 runs sampled)
Throughput for decoding: 21884.69869469849MB/s
Throughput for encoding: 2057.0803901879262MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_3.pb's performance ...

protobuf.js static decoding x 8.67 ops/sec ±1.12% (26 runs sampled)
protobuf.js static encoding x 8.02 ops/sec ±0.36% (24 runs sampled)
Throughput for decoding: 69.01108086947248MB/s
Throughput for encoding: 64.36748934222535MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_5.pb's performance ...

protobuf.js static decoding x 0.68 ops/sec ±0.30% (6 runs sampled)
protobuf.js static encoding x 1.02 ops/sec ±1.88% (7 runs sampled)
Throughput for decoding: 29.436660572775427MB/s
Throughput for encoding: 43.48865012783289MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_2.pb's performance ...

protobuf.js static decoding x 5.75 ops/sec ±11.18% (19 runs sampled)
protobuf.js static encoding x 3.55 ops/sec ±5.58% (13 runs sampled)
Throughput for decoding: 520.6985165501093MB/s
Throughput for encoding: 338.3808865696768MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_1.pb's performance ...

protobuf.js static decoding x 0.86 ops/sec ±3.01% (7 runs sampled)
protobuf.js static encoding x 0.30 ops/sec ±4.30% (5 runs sampled)
Throughput for decoding: 159.0772900797546MB/s
Throughput for encoding: 54.414657014769276MB/s
```

Our js performance:
```
Js Benchmark: 
benchmarking message benchmarks.proto2.GoogleMessage2 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/datasets/google_message2/dataset.google_message2.pb's performance ...

js deserialize x 204 ops/sec ±1.63% (85 runs sampled)
js serialize x 123 ops/sec ±0.57% (76 runs sampled)
Throughput for deserialize: 16.223843442033584MB/s
Throughput for serialize: 9.841474048590564MB/s

benchmarking message  benchmarks.proto3.GoogleMessage1 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/datasets/google_message1/proto3/dataset.google_message1_proto3.pb's performance ...

js deserialize x 103,224 ops/sec ±0.22% (88 runs sampled)
js serialize x 37,875 ops/sec ±1.47% (91 runs sampled)
Throughput for deserialize: 22.39488326426736MB/s
Throughput for serialize: 8.11572385771003MB/s

benchmarking message  benchmarks.proto2.GoogleMessage1 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/datasets/google_message1/proto2/dataset.google_message1_proto2.pb's performance ...

js deserialize x 105,884 ops/sec ±0.47% (93 runs sampled)
js serialize x 39,795 ops/sec ±0.30% (91 runs sampled)
Throughput for deserialize: 22.91463026268381MB/s
Throughput for serialize: 8.627453797348524MB/s

benchmarking message benchmarks.google_message4.GoogleMessage4 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message4.pb's performance ...

js deserialize x 0.53 ops/sec ±6.36% (6 runs sampled)
js serialize x 0.47 ops/sec ±8.94% (6 runs sampled)
Throughput for deserialize: 36.256882188294746MB/s
Throughput for serialize: 31.388194824577834MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_4.pb's performance ...

js deserialize x 140 ops/sec ±1.90% (78 runs sampled)
js serialize x 36.42 ops/sec ±2.65% (39 runs sampled)
Throughput for deserialize: 4135.529528816034MB/s
Throughput for serialize: 1070.8321706832446MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_3.pb's performance ...

js deserialize x 1.96 ops/sec ±0.59% (9 runs sampled)
js serialize x 0.91 ops/sec ±0.66% (7 runs sampled)
Throughput for deserialize: 15.666419434960796MB/s
Throughput for serialize: 7.265023049809358MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_5.pb's performance ...

js deserialize x 0.12 ops/sec ±0.95% (5 runs sampled)
js serialize x 0.10 ops/sec ±2.18% (5 runs sampled)
Throughput for deserialize: 4.998142523979117MB/s
Throughput for serialize: 4.2440661775676265MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_2.pb's performance ...

js deserialize x 0.85 ops/sec ±7.28% (7 runs sampled)
js serialize x 0.80 ops/sec ±0.90% (7 runs sampled)
Throughput for deserialize: 79.98802717792567MB/s
Throughput for serialize: 79.39654909159722MB/s

benchmarking message benchmarks.google_message3.GoogleMessage3 of dataset file /usr/local/google/home/yilunchong/mygit/protobuf/benchmarks/dataset.google_message3_1.pb's performance ...

js deserialize x 0.24 ops/sec ±5.27% (5 runs sampled)
js serialize x 0.06 ops/sec ±2.34% (5 runs sampled)
Throughput for deserialize: 44.071942529917074MB/s
Throughput for serialize: 11.191425420911754MB/s
```

What's more, our js use more memory than protobuf.js so that I have to declare more heap memory for node.js to run our js benchmark.

### CODE REVIEW

**Comment by xfxyjwf** [comment]
Thanks for putting this together!

**Comment by xfxyjwf** [comment]
In the team meeting we talked about adding closure compiler support and benchmarks on browsers. We can do that in a follow-up PR though.

File: benchmarks/js/BenchmarkSuite.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+var benchmark = require("./node_modules/benchmark");
```
Comment (xfxyjwf) [inline benchmarks/js/BenchmarkSuite.js]: I think you can just do require("benchmark")?

Comment (xfxyjwf) [inline benchmarks/js/BenchmarkSuite.js]: About the file name, should we just "benchmark_suite.js" instead?

File: benchmarks/js/JsBenchmark.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+require('./datasets/google_message1/proto2/benchmark_message1_proto2_pb.js');
```
Comment (xfxyjwf) [inline benchmarks/js/JsBenchmark.js]: Name this file "js_benchmark.js"?

Comment (xfxyjwf) [reply benchmarks/js/JsBenchmark.js]: I think we use snake_case names for JS files. Please update this file and other files in this PR as well.

**Comment by BSBandme** [comment]
Any updates for this?

File: js/binary/reader.js [hunk 3]
```diff
@@ -389,27 +389,25 @@ jspb.BinaryReader.prototype.skipFixed64Field = function() {
  */
 jspb.BinaryReader.prototype.skipGroup = function() {
   // Keep a stack of start-group tags that must be matched by end-group tags.
-  var nestedGroups = [this.nextField_];
+  var previousField = this.nextField_;
```
Comment (xfxyjwf) [inline js/binary/reader.js]: Why is this change? Can you separate it out from this pull request? Also please add tests if you are fixing a bug.

Comment (BSBandme) [reply js/binary/reader.js]: The benchmark will fail to parse if don't fix this bug.

So submit the bug fix first?

Comment (xfxyjwf) [reply js/binary/reader.js]: Yes, please do that in a separate PR with unit-tests.

**Comment by BSBandme** [comment]
Deleted skip group fixing.

**Comment by BSBandme** [comment]
Any updates for this?

**Pull request merged**